### PR TITLE
Feature: reload Lua scripts

### PR DESCRIPTION
**Pull request by nordinvannes**
Reload Lua scripts when the scripts have been changed.

### CODE REVIEW

File: scripter/lua/lua.go [hunk 1]
```diff
@@ -116,8 +149,62 @@ func (l *luaScripter) CanHandle(service string, message string) bool {
 	return false
 }
 
+// getConnIP retrieves the IP fron a connection's remote address
```
Comment (Icety) [inline scripter/lua/lua.go]: fron = from

**Review by Icety — DISMISSED** [review]
Many of the added features are now placed in lua specific files, while they do not differ for lua. I suggest therefor that they all functions that can be used for each scripting language are moved to a more global file/class.

File: scripter/lua/lua.go [hunk 2]
```diff
@@ -67,12 +71,28 @@ func (l *luaScripter) Init(service string) error {
...
 		sf := fmt.Sprintf("%s/%s/%s/%s", l.Folder, l.name, service, f.Name())
-		l.scripts[service][f.Name()] = sf
+
+		hash := ""
+		content, err := ioutil.ReadFile(sf)
+		if err == nil {
+			hasher.Reset()
+			hasher.Write(content)
+			hash = hex.EncodeToString(hasher.Sum(nil))
+		}
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: Lets make a function.

File: scripter/lua/lua.go [hunk 3]
```diff
@@ -95,7 +115,11 @@ func (l *luaScripter) GetConnection(service string, conn net.Conn) scripter.Conn
 	}
 
 	if !sConn.HasScripts(service) {
-		sConn.AddScripts(service, l.scripts[service])
+		scripts := make(map[string]string)
+		for k, v := range l.scripts[service] {
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: The variable names can be improved here, what is k and what is v?

File: scripter/scripter.go [hunk 4]
```diff
@@ -154,3 +167,75 @@ func SetBasicMethods(c ScrConn, service string) {
...
+
+// checkReloadScripts initializes services again when scripts have been changed within the service
+func checkReloadScripts(s Scripter) {
+	hasher := sha1.New()
+	for service, scripts := range s.GetScripts() {
+		isRenewService := false
+
+		// check for edited files
+		for _, script := range scripts {
+			content, err := ioutil.ReadFile(script.Source)
```
Comment (martijnjanssen) [inline scripter/scripter.go]: What happens if the file contents are pretty big? Is this going to be slow to check the files every 10s?

File: scripter/scripter.go [hunk 5]
```diff
@@ -154,3 +167,75 @@ func SetBasicMethods(c ScrConn, service string) {
 		}
 	}, service)
 }
+
+// setScriptInterval sets the interval of checking whether scripts have been changed
+func SetScriptInterval(s Scripter) {
```
Comment (martijnjanssen) [inline scripter/scripter.go]: This function does not set the interval at all. It merely enables checking the scripts. If it was going to set the interval, maybe add it as a parameter?

File: services/http.go [hunk 6]
```diff
@@ -152,6 +151,7 @@ func (s *httpService) Handle(ctx context.Context, conn net.Conn) error {
 			return err
 		}
 
+		sConn := s.scr.GetConnection("http", conn)
```
Comment (martijnjanssen) [inline services/http.go:154]: Moving this into the for-loop for the http handler doesn't make sense, what would make this nessecary?

**Review by martijnjanssen — DISMISSED** [review]
I really don't like updating the scripts on a timed interval, since by doing this, the code is probably not going to be needed 99% of the time. Even when updating the lua scripts a couple of times a day, this seems like a waste.

~Also, the checking whether there are updated scripts can be done more efficiently in my opinion. When loading them, make an array of the hashes, sorted alphabetically. When loading again, check whether the two arrays are identical, if so, the scripts are unchanged. If they are not, reload the scripts.~

File: services/http.go (hunk 6 above)
Comment (nordinvannes) [reply services/http.go:154]: This is necessary for reloading of the scripts, when a user does not leave the website, the connection is not closed. Changes made to scripts do not resolve to the end user when the connection is retrieved outside of the loop.

**Comment by nordinvannes** [comment]
In an issue we could enable script updating from the web server within Honeytrap, currently this is a side project.

File: scripter/lua/lua.go [hunk 7]
```diff
@@ -67,12 +69,23 @@ func (l *luaScripter) Init(service string) error {
...
+
 		sf := fmt.Sprintf("%s/%s/%s/%s", l.Folder, l.name, service, f.Name())
-		l.scripts[service][f.Name()] = sf
+
+		hash := ""
+		if fileStat, err := os.Stat(sf); err == nil {
+			hash = crypto.SHA1([]byte(fmt.Sprintf("%d%s", fileStat.Size(), fileStat.ModTime())))
+		}
+
+		l.scripts[service][f.Name()] = &scripter.Script{hash, sf}
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: Can't we make L77-L88 a function?

File: services/http.go (hunk 6 above)
Comment (martijnjanssen) [reply services/http.go:154]: We could change the method we reload the scripts. I think it would be better to change the luaState at the pointer location, instead of having the pointer change to a different address. This will eliminate the need of having to `GetConnection` every iteration of the for-loop.

File: scripter/abtests.json [hunk 8]
```diff
@@ -1,13 +1,23 @@
...
+  "ssh-simulator": {
     "ls": [
       "home",
       "music",
       "passwords.txt"
     ],
     "uname": [
       "Windows_NT##MS/Windows##i686##unknown##unknown##unknown##9200##6.2",
-      "Windows_NT##MS/Windows##x86_64##unknown##unknown##unknown##3790##5.2"
+      "Windows_NT##MS/Windows##x86_64##unknown##unknown##unknown##3790##5.2",
```
Comment (Icety) [inline scripter/abtests.json:10]: Shouldn't this be a separate pr, maybe even be in the lua-scripts repo?

File: scripter/lua/lua.go [hunk 9]
```diff
@@ -95,7 +108,11 @@ func (l *luaScripter) GetConnection(service string, conn net.Conn) scripter.Conn
 	}
 
 	if !sConn.HasScripts(service) {
-		sConn.AddScripts(service, l.scripts[service])
+		scripts := make(map[string]string)
```
Comment (Icety) [inline scripter/lua/lua.go]: Why not change the addScripts method to support the scripter.Script instead of translating it back here?

File: scripter/scripter.go [hunk 10]
```diff
@@ -74,6 +76,14 @@ type Result struct {
 	Content string
 }
```
Comment (Icety) [inline scripter/scripter.go]: Comments

File: scripter/scripter.go [hunk 11]
```diff
@@ -154,3 +164,14 @@ func SetBasicMethods(c ScrConn, service string) {
 		}
 	}, service)
 }
+
+// ReloadScripts reloads the scripts from the scripter
+func ReloadScripts(s Scripter) {
+	for service := range s.GetScripts() {
+		if err := s.Init(service); err != nil {
+			log.Errorf("error init service: %s", err)
```
Comment (Icety) [inline scripter/scripter.go:136]: This error message is confusing, I would make it: Error on reload, init on service failed:

File: scripter/scripter.go [hunk 12]
```diff
@@ -154,3 +164,14 @@ func SetBasicMethods(c ScrConn, service string) {
...
+		} else {
+			log.Infof("successfully updated service: %s", service)
```
Comment (Icety) [inline scripter/scripter.go:138]: Successfully reloaded service

File: server/honeytrap.go [hunk 13]
```diff
@@ -297,6 +316,21 @@ func IsTerminal(f *os.File) bool {
 	return false
 }
 
+func (hc *Honeytrap) HandleRequests(message []byte) {
```
Comment (Icety) [inline server/honeytrap.go]: Comments

File: server/honeytrap.go [hunk 14]
```diff
@@ -297,6 +316,21 @@ func IsTerminal(f *os.File) bool {
...
+	var js map[string]interface{}
+	json.Unmarshal(message, &js)
+
+	if val, ok := js["action"]; ok && val == "reload_scripts" {
+		hc.ReloadScripts()
+	}
+}
+
+func (hc *Honeytrap) ReloadScripts() {
```
Comment (Icety) [inline server/honeytrap.go]: Comments

File: services/http.go (hunk 6 above)
Comment (Icety) [reply services/http.go:154]: Why wouldn't changes to the scripts not be added to the scripts when outside of the for? They work on pointer base if I am correct and all pointers are stored in the scripter so can be called for a reload from outside.

**Review by Icety — DISMISSED** [review]

**Review by Icety — APPROVED** [review]
LGTM
